### ISSUE TITLE
"Invalid par2 files, cannot verify or repair" (1.1.0RC1), but manual unrar OK

### ISSUE THREAD

**Issue by sanderjo** [sabnzbd/sabnzbd#621]
Hi, I believe I have reported something like this before, but I can't find it:

SAB says "Invalid par2 files, cannot verify or repair" (FWIW: 1.1.0RC1) and fails.

A manual `unrar x *01.rar` (without a par2 command at all) results in the wanted, correct file. Good. But ... why can't SAB achieve that? Does SAB stumble on par2 stuff without the need for it? Seems so (see below).

EDIT: I added the NZB another time into SAB, now without par2 files, and the download goes well. So the par2 files are just an annoyance here.

So: 
- is there a good reason the `par2 r` fails ("Main packet not found.")?
- can't SAB first try a plain `unrar`, and only if that fails run par2 to repair rar files?

Details below
A manual `par2 r *001.par2` ... gives this:

```
$ par2 r *001.par2

Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol000+001.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol031+032.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol000+001.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol007+008.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol063+064.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol127+128.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol003+004.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol001+002.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol015+016.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol255+044.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol000+001.par2".
No new packets found

Main packet not found.
```
That is also in sabnzbd.log:
```
2016-07-12 08:07:38,645::INFO::[newsunpack:1338] Main packet not found...
2016-07-12 08:07:38,645::INFO::[newsunpack:1342] Extra pars = []
2016-07-12 08:07:38,646::DEBUG::[newsunpack:1551] PAR2 output was
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol127+128.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol031+032.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol000+001.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol007+008.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol063+064.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol127+128.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol127+128.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol003+004.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol001+002.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol015+016.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol255+044.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part07.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part21.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part08.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol031+032.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part04.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part03.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol000+001.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part14.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol007+008.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part19.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part02.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part15.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part18.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part12.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part06.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part16.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part17.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol063+064.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol127+128.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part11.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol003+004.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part20.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part10.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol001+002.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part09.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part13.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol015+016.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.vol255+044.par2".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part05.rar".
No new packets found
Loading "The.BlaBla.Of.S01E01.HDTV.x264-BATV.part01.rar".
No new packets found
Main packet not found.
```

That sequence is 10 times in sabnzbd.log

```
$ cat ~/.sabnzbd/logs/sabnzbd.log | grep "Main packet not found"
2016-07-12 08:05:03,049::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:05:16,881::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:05:28,166::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:05:41,213::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:06:00,324::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:06:16,827::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:06:33,287::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:06:49,657::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:07:10,434::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
2016-07-12 08:07:38,645::INFO::[newsunpack:1338] Main packet not found...
Main packet not found.
```

**Comment by Safihre** [comment]
Regarding the not just trying to continue: I think that's why we have the 'Post-Process Only Verified Jobs' switch.
If there's no par2 present at all and no SFV, it will go straight to Unpack. But it's quite a safe assumption of SABnzbd that if par2 or SFV are throwing errors that something is wrong with the files, right? :)

**Comment by sanderjo** [comment]
Ah: "Post-Process Only Verified Jobs ... Only perform post-processing on jobs that passed all PAR2 checks.".... 
So I turned it off, redownloaded the NZB, again a lot of "repairing" (WHY?), but the result is good.

> But it's quite a safe assumption of SABnzbd that if par2 or SFV are throwing errors that something is wrong with the files, right? :)

I have give you that counter proof that that is not a safe assumption ... :-)
My statement: run "unrar x" first and see what happens. That's faster than all the par2 stuff I see, even with "Post-Process Only Verified Jobs" turned off.

**Details of sabnzb.log**

It seems par2 is stilled called:
```
2016-07-12 09:19:31,609::DEBUG::[newsunpack:1551] PAR2 output was
Main packet not found.
2016-07-12 09:19:47,141::INFO::[newsunpack:1338] Main packet not found...
2016-07-12 09:19:47,142::DEBUG::[newsunpack:1551] PAR2 output was
Main packet not found.
2016-07-12 09:20:00,699::INFO::[newsunpack:1338] Main packet not found...
2016-07-12 09:20:00,700::DEBUG::[newsunpack:1551] PAR2 output was
Main packet not found.
2016-07-12 09:20:16,950::INFO::[newsunpack:1338] Main packet not found...
2016-07-12 09:20:16,952::DEBUG::[newsunpack:1551] PAR2 output was
Main packet not found.
2016-07-12 09:20:34,106::INFO::[newsunpack:1338] Main packet not found...
2016-07-12 09:20:34,108::DEBUG::[newsunpack:1551] PAR2 output was
Main packet not found.
2016-07-12 09:20:52,629::INFO::[newsunpack:1338] Main packet not found...
2016-07-12 09:20:52,630::DEBUG::[newsunpack:1551] PAR2 output was
Main packet not found.
2016-07-12 09:21:17,588::INFO::[newsunpack:1338] Main packet not found...
2016-07-12 09:21:17,589::DEBUG::[newsunpack:1551] PAR2 output was
Main packet not found.
2016-07-12 09:22:07,104::INFO::[newsunpack:1338] Main packet not found...
2016-07-12 09:22:07,105::DEBUG::[newsunpack:1551] PAR2 output was
Main packet not found.
```

**Comment by Safihre** [comment]
True, it took 3 min extra (judging from the log-times), but this is probably because the par2-files are corrupt. That's not SAB's fault.. 
If there was 1 good par2, it would have done QuickCheck using the md5-hashes (takes seconds only) and full par2 verify would also been skipped,

**Comment by thezoggy** [comment]
@Safihre sab should search for par first.. sometimes there is no visible par (nothing with extension) as they are obfuscated.. thus we should do what nzbget does.. make a quick pass to see if we can detect any.. find it and let it rename files.. THEN pp as normal. detecting a par file is pretty easy with its signature.. the problem is getting sab to do this before it pp... this was already discussed in another issue a few months back..

**Comment by Safihre** [comment]
@thezoggy this is not the case here, clearly from the logs you can see that the par2 files have normal names and are detected. They are just corrupt and can't do any renaming. 

@sanderjo the par2 files are somehow broken, so Sab is not getting the right md5 hashes from them, any of them. So it can't do QuickCheck, therefore it will force a full par2 check. 
How can it know that the files are alright, if par2 doesn't report succes and it has no other means of verification?

**Comment by sanderjo** [comment]
> How can it know that the files are alright, if par2 doesn't report succes and it has no other means of verification?

Just run `unrar x *01.rar` and check if last the line is `All OK`. If so, no need for any par2 stuff.

Just takes 5 seconds for my 400 MB .MKV file (see below). Much better than the 3 minutes back-and-forth par2-ing for no reason.

```
$ time unrar x *01.rar

UNRAR 5.40 beta 3 freeware      Copyright (c) 1993-2016 Alexander Roshal


Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part01.rar

Extracting  The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                     4%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part02.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                     9%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part03.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    14%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part04.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    19%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part05.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    24%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part06.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    29%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part07.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    34%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part08.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    39%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part09.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    44%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part10.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    49%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part11.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    54%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part12.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    59%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part13.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    63%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part14.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    68%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part15.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    73%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part16.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    78%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part17.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    83%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part18.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    88%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part19.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    93%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part20.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    98%

Extracting from The.BlaBla.Of.S01E01.HDTV.x264-BATV.part21.rar

...         The.BlaBla.Of.S01E01.HDTV.x264-BATV.mkv                    OK 
All OK

real	0m5.820s
user	0m0.728s
sys	0m0.396s
```

Ah, we got `All OK`, so we're done.

**Comment by Safihre** [comment]
I leave this one  to @shypike ;-)

**Comment by shypike** [comment]
The main reason not to attempt an unpack on an unverified set, is that this may create junk files in the form of a `__FAILED__` folder. SABnzbd doesn't clear up those kind of folders because I don't like to mess in people's media collections.
I think it's an edge case that doesn't necessarily need to be fixed.

**Comment by sanderjo** [comment]
Next attempt: `unrar t ...` , so "Test archive files". That will not create junk file, right? If the test is OK, do not bother about par files.

Reason of my next attempt: I don't think it's an edge case; it has happened a few times to me.

**Comment by shypike** [comment]
Have you looked in detail to the par2 files? Could it be that someone is using a different par tool?
I know there has been software that produced something it called par3, forgot the name of it.
"unrar t" could be an alternative verification.

**Comment by thezoggy** [comment]
wonder if this now gets handled with the obfuscation parset code....

**Comment by Safihre** [comment]
@thezoggy not better, because somehow these par2 are broken (0 blocks found in par2), also when you open them in QuickPar on Windows. Just tried it again to be sure.

**Comment by Safihre** [comment]
@sanderjo Using the new ```RarFile``` I have now implemented in #717 what you suggest. ```RarFile``` will handle the calling of ```unrar``` for us to verify.
Your test NZB will still hop from history->queue->history to get more blocks, but this is correct behavior since it is expected that if 1 par2 file is broken, the next one might be good.
In the end it will then try the new verification (if also SFV fails or doesn't exist) and proceed to unrar.
